### ISSUE TITLE
Add recommendations for detecting out-of-band installed applications in package install scripts

### ISSUE THREAD

**Issue by vexx32** [chocolatey/docs#406]
## Is Your Feature Request Related To A Problem? Please describe.

Somewhat related to the FAQ issue described in the [Automatic Sync documentation](https://docs.chocolatey.org/en-us/features/package-synchronization/automatic-sync#sync-is-not-tracking-upgrades-to-applications).

Essentially, it's not perfectly possible even in the Pro edition to keep package versions up to date with the installed software versions due to a lack of standardization and the proliferation of self-updating packages.

Some packages on the Community Repository currently have checks in place to skip the actual downloading/running of their installers if they detect the installed software version is already up to date. One example would be the `googlechrome` package. It would be good if that is provided as a recommendation, as it will help users avoid unnecessary reinstalls or errors when doing `choco upgrade all` for example.

## Describe The Solution. Why is it needed?

We should add some recommendations to the [packaging documentation](https://docs.chocolatey.org/en-us/create/create-packages) that maintainers creating or updating their packages, especially those handling software that is capable of auto-updating itself, include some mechanism to detect whether the software was updated out-of-band. In this case, if the software is detected as being already up to date, the install script can simply `return` early instead of re-running the installer and potentially modifying a perfectly good installation.

I would also suggest packages check if the package is being installed with `-f` / `--force` which appears to be passed into the process environment as `$env:chocolateyForce` (see environment code [here](https://github.com/chocolatey/choco/blob/ababf506acde28b6dd24b3183af1e28a8ef6ebe1/src/chocolatey/infrastructure.app/configuration/EnvironmentSettings.cs#L130)) -- though an argument could be made that a lot of these aren't documented and probably should be, but that's a separate issue, just mentioning this here as this one is particularly relevant.

My thought is if users do `choco install <package> --force` then the package should look to detect already-installed versions and do a full reinstall in that case. I don't know if that's a sensible suggestion for every use case, so I'm happy to discuss on that point for sure.

## Related Issues

N/A

**Comment by TheCakeIsNaOH** [comment]
> In this case, if the software is detected as being already up to date, the install script can simply return early instead of re-running the installer and potentially modifying a perfectly good installation.

The check should specifically should be if the software version from the package is equal to the currently installed version. Because there are cases where a user may install an older version intentionally, or use `--allow-downgrade` to move to a lower version.

>  though an argument could be made that a lot of these aren't documented and probably should be,

`$env:ChocolateyForce` is documented here: https://docs.chocolatey.org/en-us/create/functions/#set-by-options-and-configuration

I do agree that they could be documented better.

**Comment by pauby** [comment]
> full reinstall

What do you mean by this?

I feel that when reading this we need to remember there is a steep learning curve in creating packages and that the vast majority of maintainers do this in their free time. If we start to request that they add additional checks in packages, we're going to make that pool of maintainers who are prepared to do this smaller. And reduce the number of maintainers who come on board smaller due to the higher learning curve.

I know that we are not talking about _requiring_ any of this here, but when you start to add it to documentation, you need to be careful that it doesn't come across that way. I'd suggest, if we are going to do this, it's not added to the Create Packages page, but somewhere else, perhaps for 'Advanced Packaging'.

I also feel that we are making this issue bigger than it is. There are users who have complained about packages, but I don't see many of those complaints and I feel they are better addressed on the individual packages. 

I am currently working on, in my free time, videos to help to create and maintain packages which I think would be useful here.

**Comment by vexx32** [comment]
Yeah for sure, I don't think it's worth requiring anything. I think the main thing is having visible suggestions for what end-users of packages may find useful. It's something that Chocolatey itself can't really enforce in a meaningful way because of how different installers in the Windows ecosystem tend to be, so I think it might be useful to give maintainers a shortlist of "here's some stuff that might be worth looking into" for packages.

I don't think it needs to be hugely prescriptive or demanding, just a few things like this which can improve the overall UX for a package with perhaps some links to (or code snippets from) a few packages that make a good example case so folks that want to provide things like this have the tools to do so a bit more easily, and we have a place to point folks for some help.

Videos would be a great addition as well, absolutely.
